### PR TITLE
Dev reg tests

### PR DESCRIPTION
**Pull request by FrauDeborah**
Hi Johnathan (@haxwell )

**devRegTests** is my first attempt  at writing tests of the register.service for the registration component. Or more accurately, the first I've pushed for you to review. I had many false starts.

I used injection and spyOn() to test for success and for failure of the three methods (isUniqueUsername, isUniqueEmail, and registerUser) which call the api.service. I'm hoping the 6 tests are written correctly. All ng tests ran successfully, ng lint results corrected, and ng serve compiled and ran.

Feel free to merge if the tests are worthy. If not, any feedback is appreciated.

Deborah

PS. When I took a break late Sunday afternoon, I fell asleep awaking 8 hours later. Wow, I must have needed the sleep! Anyway, my feeble reason for missing the Sunday evening pull request suspense. "O'Dark-Thrity" on Monday morning is likely just as good. See you at the Sprint Planning meeting this evening.

### CODE REVIEW

File: src/app/registration/register.service.spec.ts [hunk 1]
```diff
@@ -2,18 +2,61 @@ import { TestBed, inject } from '@angular/core/testing';
...
        ]
     });
   });
 
   it('should be created', inject([RegisterService], (service: RegisterService) => {
     expect(service).toBeTruthy();
   }));
+
+  it('isUniqueUsername(uniqueUsername) should be truthy', inject([RegisterService], (service: RegisterService) => {
+    spyOn(service, 'isUniqueUsername').and.returnValue(Promise.resolve(true));
```
Comment (haxwell) [inline src/app/registration/register.service.spec.ts]: First, very nice to see you using Spies and all that good stuff.. This is where you begin to get into the meat of unit testing..

What you want to do with this tests, is verify that a method operates as you would expect. So, for the method isUniqueUsername(), we are able to follow the following steps in its code:

- Well, there's only one, this is a wrapper for an API call, who's details are hidden to us. Our One Step, is to return the value from this API.

The point of a Spy is to allow us to define what happens when a method is called. This is useful because when we are testing, we can create our own conditions and verify what happens when a method is called.

In this case, we want to verify what happens when the `post()` API is called. So, we set our Spy, not on the method we are testing, but on the method who's behavior we want to control.

```
it('is a test', inject([RegisterService, ApiService], (service: RegisterService, apiService: ApiService) => {
    spyOn(apiService, 'post').and.returnValue(Promise.resolve(true));
...
}
```

And then, as you have it already, we would check to see if the method we are testing, `service.isUniqueUsername` returns true-ish, as we would expect.

But what you have will not work, because `apiService` is returning a Promise; it is asynchronous, and the code you've written assumes it is not. See this answer on StackOverflow: https://stackoverflow.com/a/35431221/512561

**Comment by haxwell** [comment]
No worries on missing the "deadline"... the effort you put in is reassurance enough. Thank you.

File: src/app/registration/registration.component.spec.ts [hunk 2]
```diff
@@ -1,14 +1,14 @@
 import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
 import { async, ComponentFixture, TestBed } from '@angular/core/testing';
 
-import { 
+import {
```
Comment (haxwell) [inline src/app/registration/registration.component.spec.ts:4]: Files like this, and a couple others in this PR, which do not change any functionality, but only do minor whitespace-type changes, should be in the own PR. In this case, it is fine, just something to keep in mind. Once PRs get big, it can easy to overlook stuff when unrelated changes like this are in the mix.

Comment (FrauDeborah) [reply src/app/registration/registration.component.spec.ts:4]: Ok will do in the future!

File: src/app/registration/register.service.spec.ts (hunk 1 above)
Comment (FrauDeborah) [reply src/app/registration/register.service.spec.ts]: Hi Johnathan (@haxwell),

I revised the register service tests on the devRegTests branch per your guidance above. Yet, I'm still not sure I have the revised tests correctly. Also,  what should afterEach() do to remove the injected services. Or can afterEach() be eliminated for these tests.

Feedback would be appreciated. 

Deborah

**Comment by FrauDeborah** [comment]
Depending upon the status of the PRs "Dev Fix Login" and "DevFL Check Api Routing". I may be able to close this PR. Just waiting a few more days tp verify status.
